xapian-applications/omega/docs/newformat.rst

```diff
@@ -0,0 +1,174 @@
+
+   The ideal (and simplest) case is that you have a filter which can produce an UTF-8 output in plain text. It may requiere special command line options to do so, in which case work out what they are from the documentation or source code, and check that the output is indeed as documented.
+
+   It is most efficient if the filter program can write to stdout, but output to a temporary file works too.
+
+   For example, if we want to use ``python2text`` for handling ``text/x-python``, we should use ``python2text --utf8 --stdout``.
+
+2. Then, we need to add the filter to Omega. Omega has the ability to specify additional external filters on the command line using ``--filter=M[,[T][,C]]:CMD``, which process files with MIME Content-Type M through command CMD and produces output (on stdout or in a temporary file) with format T (Content-Type or file extension; currently txt (default), html or svg) in character encoding C (default: UTF-8). For example
+   ::
+
+     --filter=text/x-foo,text/html,utf-16:'foo2utf16 --content %f %t'
+
+   In this case, we are going to handle ``text/x-foo`` files with ``foo2utf16`` that is going to produce html with UTF-16 encoding on a temporary file. Note that %f will be replaced with the filename and %t with a temporary output file (that is going to be created by omindex at runtime). This tells omindex to index files with content-type ``text/x-foo`` by running
+   ::
+
+     foo2utf16 --content path/to/file path/to/temporary/file
+
+   If we want to add the filter permanently, we can add a new entry in ``index_add_default_filters`` at 'index_file.cc'. Following with the example
+   ::
+
+     index_command(“text/x-foo”, Filter(“foo2utf16 --content %f %t”, "text/html", “utf-16”))
+
+   There are more options that we can use for Filter (see 'index_file.h').
+
+3. In some cases, you will have to run several programs for each file or make some extra work so you will either need to put together a script which fits what omindex supports, or else modify the source code in ‘index_file.cc’ by adding a test for the new mime-type to the long if/else-if chain inside ``index_mimetype`` function. New formats should generally go at the end, unless they are very common
+   ::
+
+     } else if (mimetype == "text/x-foo") {
+
+   The filename of the file is in ``file``. The code you add should at least extract the "body" text of the document into the C++ variable ``dump``. Optionally, you can also set ``title`` (the document's title), ``keywords`` (additional text to index, but not to show the user), ``sample`` (if set, this is used to generate the static document "snippet" which is stored; if not set, this is generated from dump) and ``topic``
+   ::
+
+     string tmpfile = get_tmpfile("tmp.html");
+     if (tmpfile.empty())
+       return;
+     string safetmp = shell_protect(tmpfile);
+     string cmd = "foo2utf16 --content " + shell_protect(file) + " " + safetmp;
```

Comment:
This code snippet is rather out of date - `shell_protect()` was replaced in 2012.

This should now be something like:

```
string cmd = "foo2utf16 --content";
append_filename_argument(cmd, file);
append_filename_argument(cmd, tmpfile);
```

Note `append_filename_argument` appends a leading space for you (unless told not to).

Reply:
Sorry, I didn't realize. I am going to update it.